From adfe96867460409750633e17e12e0e5146a87b2a Mon Sep 17 00:00:00 2001
From: Alex James <git@alextjam.es>
Date: Mon, 28 Apr 2025 00:20:08 -0500
Subject: [PATCH 1/3] glob: avoid deleting non-static d_share member

Clang warns that this is UB:

> glob5.cc:28:12: error: cannot refer to a non-static member from the handler of a constructor function try block [-Werror,-Wexceptions]
>    28 |     delete d_share;
>       |            ^
---
 support/bobcat/glob/glob5.cc | 45 +++++++++++++++--------------
 1 file changed, 24 insertions(+), 21 deletions(-)

diff --git a/support/bobcat/glob/glob5.cc b/support/bobcat/glob/glob5.cc
index 3d70063..90b09dc 100644
--- a/support/bobcat/glob/glob5.cc
+++ b/support/bobcat/glob/glob5.cc
@@ -2,29 +2,32 @@
 
 Glob::Glob(unordered_set<Type> const &type, string const &pattern,
            int flags, Dots dots)
-try
-:
-    d_share(new GlobShare { glob_t{}, 1, type } )
 {
-    if (flags & ~Flags::mask)
-        throw Exception(flags) << "Glob: unknown Flag specified";
+    auto share = new GlobShare { glob_t{}, 1, type };
+    d_share = share;
 
-    int err = glob(pattern.c_str(), flags & ~NOMATCH, 0,
-                                                    &d_share->globStruct);
-
-    if (err != 0)
+    try
     {
-        if (not (err == GLOB_NOMATCH and flags & NOMATCH))
-            throw Exception(err) << "Glob: glob() failed";
+        if (flags & ~Flags::mask)
+            throw Exception(flags) << "Glob: unknown Flag specified";
+
+        int err = glob(pattern.c_str(), flags & ~NOMATCH, 0,
+                                                        &d_share->globStruct);
+
+        if (err != 0)
+        {
+            if (not (err == GLOB_NOMATCH and flags & NOMATCH))
+                throw Exception(err) << "Glob: glob() failed";
+        }
+
+        accept();
+
+        if (dots == FIRST)
+            stable_partition(mbegin(), mend(), isDot);
+    }
+    catch (...)
+    {
+        delete share;
+        throw;
     }
-
-    accept();
-
-    if (dots == FIRST)
-        stable_partition(mbegin(), mend(), isDot);
-}
-catch(...)
-{
-    delete d_share;
-    throw;
 }
-- 
2.48.1

